{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "473e7da4-ed9a-45ce-8335-8f6f3f6dd225",
   "metadata": {},
   "outputs": [],
   "source": [
    "def return_svg(svg: str):\n",
    "    with open(svg, \"r\") as f:\n",
    "        return f.read()\n",
    "    \n",
    "def create_report(plot,\n",
    "                  plot2,\n",
    "                  pdf, \n",
    "                  svg\n",
    ") -> None:\n",
    "    \"\"\"\n",
    "    creates html report\n",
    "    \"\"\"\n",
    "    \n",
    "    html = f\"\"\"\n",
    "    <!DOCTYPE html>\n",
    "    <html>\n",
    "        <head>\n",
    "            <title>Report</title>\n",
    "            <link href=\"https://cdn.jsdelivr.net/npm/@mdi/font@latest/css/materialdesignicons.min.css\" rel=\"stylesheet\">\n",
    "            <script src=\"https://cdn.jsdelivr.net/npm/vega@5\"></script>\n",
    "            <script src=\"https://cdn.jsdelivr.net/npm/vega-lite@5\"></script>\n",
    "            <script src=\"https://cdn.jsdelivr.net/npm/vega-embed@6\"></script>\n",
    "        </head>\n",
    "        <body>\n",
    "            <h1 style=\"color:blue; text-align:center; font-style:italic;\">\n",
    "                Welcome to The report of the sample! \n",
    "            </h1>\n",
    "            <!-- PLOT1 -->\n",
    "            <h3 style=\"font-style:italic; text-align:center;\"> \n",
    "                This is a plot of a beautiful graph I just made! Enjoy!\n",
    "            </h3>\n",
    "            \n",
    "            <div id=\"vis\" style=\"display:flex;justify-content:center;align-items:center;width:100%;height:100%;\"></div>\n",
    "            <script type=\"text/javascript\">\n",
    "                vegaEmbed(\"#vis\", {plot});\n",
    "            </script>\n",
    "            \n",
    "            <h1 style=\"color:red;\"> \n",
    "                Another plot \n",
    "            </h1>\n",
    "            <div id=\"vis2\" style=\"display:flex;justify-content:center;align-items:center;width:100%;height:100%;\"></div>\n",
    "            <script type=\"text/javascript\">\n",
    "                vegaEmbed(\"#vis2\", {plot2});\n",
    "            </script>\n",
    "            \n",
    "            <h1> \n",
    "                This is a PDF \n",
    "            </h1>\n",
    "            <!-- <iframe src=\"{pdf}#toolbar=0\" width=\"100%\" height=\"900px\" border=\"0\"> -->\n",
    "            </iframe>\n",
    "            \n",
    "            <h1 style=\"color:orange; text-align:center; font-style:italic;\">\n",
    "                This is an embedded SVG\n",
    "            </h1>\n",
    "            <div style=\"display:flex;justify-content:center;align-items:center;width:100%;height:100%;\">\n",
    "                {svg}\n",
    "            </div>\n",
    "            \n",
    "        </body>\n",
    "    </html>\n",
    "    \"\"\"\n",
    "    \n",
    "    with open(\"hej.html\", \"w\") as f:\n",
    "        print(html, file=f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8569fc6b-591d-4344-96cb-643b0001cee1",
   "metadata": {},
   "source": [
    "# Plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4ca37dc8-ebea-43e6-a782-02c7802b6c10",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import plotting functions from plotting\n",
    "from plotting import bracken_raw, contig_quality, kaiju_raw, kaiju_megahit, cat_megahit\n",
    "from pathlib import Path\n",
    "import altair as alt\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cbfab24-50df-48f2-a1b9-7bbfa9d5e536",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "bb12a994-8605-4951-a760-3dcbf33ac10f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'sample11_S6'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Samples\n",
    "SAMPLES = [x for x in Path(\"../virusHanter/static/data\").iterdir() if x.is_dir() and not x.stem.startswith(\".\")]\n",
    "sample11 = Path(\"../virusclassification_nextflow/results/sample11_S6/\")\n",
    "sample11.parts[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "28945ab9-fcda-49a5-8f48-d81494d777c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>percent</th>\n",
       "      <th>reads</th>\n",
       "      <th>taxon_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10298</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>0.056239</td>\n",
       "      <td>128</td>\n",
       "      <td>Human alphaherpesvirus 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10298</td>\n",
       "      <td>Duplodnaviria</td>\n",
       "      <td>0.056239</td>\n",
       "      <td>128</td>\n",
       "      <td>Human alphaherpesvirus 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10298</td>\n",
       "      <td>Human alphaherpesvirus 1</td>\n",
       "      <td>0.056239</td>\n",
       "      <td>128</td>\n",
       "      <td>Human alphaherpesvirus 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10298</td>\n",
       "      <td>Simplexvirus</td>\n",
       "      <td>0.056239</td>\n",
       "      <td>128</td>\n",
       "      <td>Human alphaherpesvirus 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10298</td>\n",
       "      <td>Alphaherpesvirinae</td>\n",
       "      <td>0.056239</td>\n",
       "      <td>128</td>\n",
       "      <td>Human alphaherpesvirus 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>332937</td>\n",
       "      <td>Herviviricetes</td>\n",
       "      <td>0.000439</td>\n",
       "      <td>1</td>\n",
       "      <td>Chimpanzee herpesvirus strain 105640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>332937</td>\n",
       "      <td>Herpesvirales</td>\n",
       "      <td>0.000439</td>\n",
       "      <td>1</td>\n",
       "      <td>Chimpanzee herpesvirus strain 105640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>332937</td>\n",
       "      <td>Herpesviridae</td>\n",
       "      <td>0.000439</td>\n",
       "      <td>1</td>\n",
       "      <td>Chimpanzee herpesvirus strain 105640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>332937</td>\n",
       "      <td>Alphaherpesvirinae</td>\n",
       "      <td>0.000439</td>\n",
       "      <td>1</td>\n",
       "      <td>Chimpanzee herpesvirus strain 105640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>1980494</td>\n",
       "      <td>Tula orthohantavirus</td>\n",
       "      <td>0.000439</td>\n",
       "      <td>1</td>\n",
       "      <td>Tula orthohantavirus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>97 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    taxon_id                  taxonomy   percent  reads  \\\n",
       "0      10298                   Viruses  0.056239    128   \n",
       "1      10298             Duplodnaviria  0.056239    128   \n",
       "2      10298  Human alphaherpesvirus 1  0.056239    128   \n",
       "3      10298              Simplexvirus  0.056239    128   \n",
       "4      10298        Alphaherpesvirinae  0.056239    128   \n",
       "..       ...                       ...       ...    ...   \n",
       "92    332937            Herviviricetes  0.000439      1   \n",
       "93    332937             Herpesvirales  0.000439      1   \n",
       "94    332937             Herpesviridae  0.000439      1   \n",
       "95    332937        Alphaherpesvirinae  0.000439      1   \n",
       "96   1980494      Tula orthohantavirus  0.000439      1   \n",
       "\n",
       "                              taxon_name  \n",
       "0               Human alphaherpesvirus 1  \n",
       "1               Human alphaherpesvirus 1  \n",
       "2               Human alphaherpesvirus 1  \n",
       "3               Human alphaherpesvirus 1  \n",
       "4               Human alphaherpesvirus 1  \n",
       "..                                   ...  \n",
       "92  Chimpanzee herpesvirus strain 105640  \n",
       "93  Chimpanzee herpesvirus strain 105640  \n",
       "94  Chimpanzee herpesvirus strain 105640  \n",
       "95  Chimpanzee herpesvirus strain 105640  \n",
       "96                  Tula orthohantavirus  \n",
       "\n",
       "[97 rows x 5 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cleaned_bracken_report = list(sample11.rglob(\"*bracken_raw.csv\"))[0]\n",
    "cleaned_kaiju_report = list(sample11.rglob(\"*kaiju_raw.csv\"))[0]\n",
    "\n",
    "\n",
    "number = 10\n",
    "kaiju = pd.read_csv(cleaned_kaiju_report)\n",
    "kaiju"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3a326fe6-296f-4d14-a67d-b27b7eb3b9ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-eb9a964396cb448b9f939f78b6d20b34\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-eb9a964396cb448b9f939f78b6d20b34\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-eb9a964396cb448b9f939f78b6d20b34\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.17.0?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"4.17.0\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}}, \"data\": {\"name\": \"data-b0c4e0d3bc9885e34157763d54e37ead\"}, \"mark\": \"bar\", \"encoding\": {\"color\": {\"field\": \"name\", \"title\": null, \"type\": \"nominal\"}, \"tooltip\": [{\"field\": \"domain\", \"type\": \"nominal\"}, {\"field\": \"new_est_reads\", \"title\": \"Number of reads\", \"type\": \"quantitative\"}], \"x\": {\"axis\": {\"format\": \".1%\"}, \"field\": \"percent\", \"title\": \"Percent of reads\", \"type\": \"quantitative\"}, \"y\": {\"field\": \"name\", \"sort\": \"-x\", \"title\": null, \"type\": \"nominal\"}}, \"height\": 500, \"title\": \"Bracken classification raw\", \"width\": 500, \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.17.0.json\", \"datasets\": {\"data-b0c4e0d3bc9885e34157763d54e37ead\": [{\"percent\": 0.1236999999999999, \"new_est_reads\": 259, \"level\": \"S\", \"tax_id\": 10298, \"name\": \"                    Human alphaherpesvirus 1\", \"domain\": \"Virus\"}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bracken_bar_plot = bracken_raw.bar_chart_bracken_raw(\n",
    "    cleaned_bracken_report, number=number,virus_only=True\n",
    ")\n",
    "bracken_bar_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d70681f6-ff2e-40b0-85d7-99308f68b985",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-d3853b14295c4e43aa91babc93e05d24\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-d3853b14295c4e43aa91babc93e05d24\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-d3853b14295c4e43aa91babc93e05d24\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.17.0?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"4.17.0\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}}, \"data\": {\"name\": \"data-5c5887267d2503c7bbab9478104daf29\"}, \"mark\": \"bar\", \"encoding\": {\"color\": {\"field\": \"taxon_name\", \"title\": null, \"type\": \"nominal\"}, \"tooltip\": [{\"field\": \"taxonomy\", \"type\": \"ordinal\"}, {\"field\": \"reads\", \"title\": \"Number of reads\", \"type\": \"quantitative\"}], \"x\": {\"axis\": {\"format\": \".1%\"}, \"field\": \"percent\", \"title\": \"Percent of reads\", \"type\": \"quantitative\"}, \"y\": {\"field\": \"taxon_name\", \"sort\": \"-x\", \"title\": null, \"type\": \"nominal\"}}, \"height\": 500, \"title\": \"Kaiju classification raw\", \"width\": 500, \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.17.0.json\", \"datasets\": {\"data-5c5887267d2503c7bbab9478104daf29\": [{\"taxon_id\": 10298, \"percent\": 0.05623902, \"taxon_name\": \"Human alphaherpesvirus 1\", \"reads\": 128, \"taxonomy\": [\"Viruses\", \"Duplodnaviria\", \"Human alphaherpesvirus 1\", \"Simplexvirus\", \"Alphaherpesvirinae\", \"Herpesviridae\", \"Herpesvirales\", \"Herviviricetes\", \"Peploviricota\", \"Heunggongvirae\"]}, {\"taxon_id\": 10294, \"percent\": 0.01669596, \"taxon_name\": \"Simplexvirus\", \"reads\": 38, \"taxonomy\": [\"Herpesvirales\", \"Peploviricota\", \"Herviviricetes\", \"Viruses\", \"Herpesviridae\", \"Alphaherpesvirinae\", \"Duplodnaviria\", \"Simplexvirus\", \"Heunggongvirae\"]}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bracken_bar_plot = bracken_raw.bar_chart_bracken_raw(\n",
    "    cleaned_bracken_report, number=number,virus_only=True\n",
    ")\n",
    "\n",
    "bracken_domain_bar_plot = bracken_raw.bar_chart_bracken_raw(\n",
    "    cleaned_bracken_report, level=\"domain\", virus_only=False\n",
    ")\n",
    "\n",
    "kaiju_bar_plot = kaiju_raw.bar_chart_kaiju_raw(file=cleaned_kaiju_report)\n",
    "\n",
    "species_and_domain_bracken = (\n",
    "    alt.hconcat(bracken_bar_plot, bracken_domain_bar_plot)\n",
    "    .resolve_scale(color=\"independent\")\n",
    ")\n",
    "\n",
    "species_and_domain_bracken\n",
    "kaiju_bar_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8d031c57-e390-4ade-9b88-e3147f0da83b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_report(\n",
    "    bracken_report,\n",
    "    bracken_report2,\n",
    "    sample,\n",
    ") -> None:\n",
    "    \"\"\"\n",
    "    Creates html report\n",
    "    \"\"\"\n",
    "    html = f\"\"\"\n",
    "    <!DOCTYPE html>\n",
    "    <html>\n",
    "        <head>\n",
    "            <meta charset=\"utf-8\">\n",
    "            <title>Report</title>\n",
    "            <link href=\"https://cdn.jsdelivr.net/npm/@mdi/font@latest/css/materialdesignicons.min.css\" rel=\"stylesheet\">\n",
    "            <script src=\"https://cdn.jsdelivr.net/npm/vega@5\"></script>\n",
    "            <script src=\"https://cdn.jsdelivr.net/npm/vega-lite@5\"></script>\n",
    "            <script src=\"https://cdn.jsdelivr.net/npm/vega-embed@6\"></script>\n",
    "            <style>\n",
    "                body {{\n",
    "                font-family: sans-serif;\n",
    "                }}\n",
    "                header {{\n",
    "                background-color: #333;\n",
    "                color: white;\n",
    "                padding: 20px;\n",
    "                text-align: center;\n",
    "                }}\n",
    "                section {{\n",
    "                display: flex;\n",
    "                min-height: 100vh; /* Make the section take up the full height of the page */\n",
    "                }}\n",
    "                aside {{\n",
    "                width: 200px; /* Fixed width for the sidebar */\n",
    "                background-color: #ccc;\n",
    "                padding: 20px;\n",
    "                font-family: sans-serif;\n",
    "                font-size: 14px;\n",
    "                border-radius: 5px;\n",
    "                box-shadow: 2px 2px 5px rgba(0, 0, 0, 0.3);\n",
    "                }}\n",
    "                main {{\n",
    "                flex: 1; /* Take up remaining space */\n",
    "                padding: 20px;\n",
    "                }}\n",
    "                h1 {{\n",
    "                margin-top: 0;\n",
    "                }}\n",
    "                h2 {{\n",
    "                margin-top: 0;\n",
    "                margin-bottom: 10px;\n",
    "                }}\n",
    "                ul {{\n",
    "                list-style: none;\n",
    "                margin: 0;\n",
    "                padding: 0;\n",
    "                }}\n",
    "                li {{\n",
    "                margin: 5px 0;\n",
    "                }}\n",
    "                a {{\n",
    "                color: #333;\n",
    "                text-decoration: none;\n",
    "                }}\n",
    "                a:hover {{\n",
    "                text-decoration: underline;\n",
    "                }}\n",
    "            </style>\n",
    "        </head>\n",
    "        <body>\n",
    "            <header>\n",
    "                <h1>Report</h1>\n",
    "            </header>\n",
    "            <section>\n",
    "                <aside>\n",
    "                    <h2>Links</h2>\n",
    "                    <ul>\n",
    "                        <li><a href=\"#\">Link 1</a></li>\n",
    "                        <li><a href=\"#\">Link 2</a></li>\n",
    "                        <li><a href=\"#\">Link 3</a></li>\n",
    "                    </ul>\n",
    "                </aside>\n",
    "                <main>\n",
    "                    <h1 style=\"color:blue; text-align:center; font-style:italic;\">\n",
    "                        Welcome to The report of the sample! \n",
    "                    </h1>\n",
    "                    <!-- PLOT1 -->\n",
    "                    <div id=\"bracken_report\" style=\"display:flex;justify-content:center;align-items:center;width:100%;height:100%;\"></div>\n",
    "                    <script type=\"text/javascript\">\n",
    "                        vegaEmbed(\"#bracken_report\", {bracken_report});\n",
    "                    </script>\n",
    "                    <!-- PLOT2 -->\n",
    "                    <div id=\"bracken_report2\" style=\"display:flex;justify-content:center;align-items:center;width:100%;height:100%;\"></div>\n",
    "                    <script type=\"text/javascript\">\n",
    "                        vegaEmbed(\"#bracken_report2\", {bracken_report2});\n",
    "                    </script>\n",
    "                </main>\n",
    "            </section>\n",
    "        </body>\n",
    "    </html>\n",
    "    \"\"\"\n",
    "    \n",
    "    with open(f\"{sample}_report.html\", \"w\") as f:\n",
    "        print(html, file=f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a91da918-87c9-4932-a65b-35392f2ffafa",
   "metadata": {},
   "outputs": [],
   "source": [
    "create_report(bracken_bar_plot.to_json(), bracken_bar_plot.to_json(), sample=\"test\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "74d336c4-28d6-4736-93e8-cc32d532e853",
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to create the report\n",
    "def create_report(\n",
    "    sample_name: str,\n",
    "    plot: str,\n",
    "    svg: str,\n",
    ") -> None:\n",
    "    \"\"\"\n",
    "    Creates html report\n",
    "    \"\"\"\n",
    "  \n",
    "    html = f\"\"\"\n",
    "    <!DOCTYPE html>\n",
    "    <html>\n",
    "        <head>\n",
    "            <title>Report</title>\n",
    "            <link href=\"https://cdn.jsdelivr.net/npm/@mdi/font@latest/css/materialdesignicons.min.css\" rel=\"stylesheet\">\n",
    "            <script src=\"https://cdn.jsdelivr.net/npm/vega@5\"></script>\n",
    "            <script src=\"https://cdn.jsdelivr.net/npm/vega-lite@5\"></script>\n",
    "            <script src=\"https://cdn.jsdelivr.net/npm/vega-embed@6\"></script>\n",
    "            <style>\n",
    "            header {{\n",
    "            background-color: #333;\n",
    "            color: white;\n",
    "            padding: 20px;\n",
    "            text-align: center;\n",
    "            }}\n",
    "            </style>\n",
    "        </head>\n",
    "        <body>\n",
    "            <header>\n",
    "                <h1>Report</h1>\n",
    "            </header>\n",
    "            <h1 style=\"color:blue; text-align:center; font-style:italic;\">\n",
    "                Welcome to The report of the sample! \n",
    "            </h1>\n",
    "            <!-- PLOT1 -->\n",
    "            <h3 style=\"font-style:italic; text-align:center;\"> \n",
    "                This is a plot of a beautiful graph I just made! Enjoy!\n",
    "            </h3>\n",
    "            \n",
    "            <div id=\"vis\" style=\"display:flex;justify-content:center;align-items:center;width:100%;height:100%;\"></div>\n",
    "            <script type=\"text/javascript\">\n",
    "                vegaEmbed(\"#vis\", {plot});\n",
    "            </script>\n",
    "            \n",
    "            <h1> \n",
    "                This is a PDF \n",
    "            </h1>\n",
    "            <!-- <iframe src=\"{{pdf}}#toolbar=0\" width=\"100%\" height=\"900px\" border=\"0\"> -->\n",
    "            </iframe>\n",
    "            \n",
    "            <h1 style=\"color:orange; text-align:center; font-style:italic;\">\n",
    "                This is an embedded SVG\n",
    "            </h1>\n",
    "            <div style=\"display:flex;justify-content:center;align-items:center;width:100%;height:100%;\">\n",
    "                {svg}\n",
    "            </div>\n",
    "            \n",
    "        </body>\n",
    "    </html>\n",
    "    \"\"\"\n",
    "    \n",
    "    with open(f\"{sample_name}_report.html\", \"w\") as f:\n",
    "        print(html, file=f)\n",
    "        \n",
    "svg = return_svg(\"visualization (1).svg\")\n",
    "create_report(\"test\", bracken_bar_plot.to_json(), svg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "11be06e0-7bc6-464c-92b3-f05ab756dae1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "def parse_alignments(log_file: str) -> tuple[int, float]:\n",
    "    \"\"\"Parses out the total number of reads and the percentage of reads that were aligned to the reference genome from a bowtie2 log file.\n",
    "    \n",
    "    Args:\n",
    "        log_file: The path to the log file.\n",
    "    \n",
    "    Returns:\n",
    "        A tuple containing the total number of reads as an int and the percentage of reads that were aligned to the\n",
    "        reference genome as a float, or None if the percentage could not be parsed.\n",
    "    \"\"\"\n",
    "    total_reads = 0\n",
    "    percent_aligned = None\n",
    "    pattern_total = r\"^(\\d+) reads; of these:$\"\n",
    "    pattern_aligned = r\"^[\\d.]+\\% overall alignment rate$\"\n",
    "    with open(log_file) as f:\n",
    "        for line in f:\n",
    "            match = re.search(pattern_total, line)\n",
    "            if match:\n",
    "                total_reads = int(match.group(1))\n",
    "            match = re.search(pattern_aligned, line)\n",
    "            if match:\n",
    "                percent_aligned = float(match.group(0).split(\"%\")[0])\n",
    "    return total_reads, percent_aligned\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0af1b106-bc16-4cd9-83c1-17f1ec2a0ce8",
   "metadata": {},
   "outputs": [],
   "source": [
    "log = \"../virusHanter/static/data/sample09_S8/bowtie2/sample09_S8.bowtie_raw.log\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "266c0fa6-1cc2-478f-b02c-adb68c93f498",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import altair as alt\n",
    "number_aligned = int(total_reads * percent_aligned / 100)\n",
    "number_unaligned = total_reads - number_aligned\n",
    "\n",
    "\n",
    "bowtie_df = pd.DataFrame({\"amount\": [number_unaligned, number_aligned],\n",
    "                         \"type\": [\"unaligned\", \"aligned\"]})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8dda0455-8d50-40db-ade0-e96ee68ce8ed",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>amount</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8177</td>\n",
       "      <td>aligned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>193250</td>\n",
       "      <td>unaligned</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   amount       type\n",
       "0    8177    aligned\n",
       "1  193250  unaligned"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bowtie_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "35c329b5-bc1d-4adf-821f-fb9d2fca6532",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-456b02e546a848598913fe60c7372f27\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-456b02e546a848598913fe60c7372f27\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-456b02e546a848598913fe60c7372f27\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.17.0?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"4.17.0\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}}, \"data\": {\"name\": \"data-eb3ee5c90b71e8cc55c3c1d229156cbe\"}, \"mark\": \"bar\", \"encoding\": {\"color\": {\"field\": \"type\", \"type\": \"nominal\"}, \"tooltip\": [{\"field\": \"amount\", \"title\": \"Number of reads aligned\", \"type\": \"quantitative\"}, {\"field\": \"type\", \"type\": \"nominal\"}], \"x\": {\"aggregate\": \"sum\", \"axis\": {\"format\": \"%\"}, \"field\": \"amount\", \"stack\": \"normalize\", \"title\": \"Percent\", \"type\": \"quantitative\"}}, \"title\": \"Reads aligned to the human reference genome\", \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.17.0.json\", \"datasets\": {\"data-eb3ee5c90b71e8cc55c3c1d229156cbe\": [{\"amount\": 193250, \"type\": \"unaligned\"}, {\"amount\": 8177, \"type\": \"aligned\"}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alt.Chart(bowtie_df).mark_bar().encode(\n",
    "    x=alt.X('sum(amount)', \n",
    "            stack=\"normalize\", \n",
    "            axis=alt.Axis(format='%'), \n",
    "            title=\"Percent\"),\n",
    "    color='type',\n",
    "    tooltip=[alt.Tooltip(\"amount:Q\", title=\"Number of reads aligned\"),\n",
    "             alt.Tooltip(\"type:N\")]\n",
    ").properties(title=\"Reads aligned to the human reference genome\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f1fc994c-26fc-4be7-8266-d0772e62a222",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-2c883b498f32429d984e65263ac7348a\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-2c883b498f32429d984e65263ac7348a\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-2c883b498f32429d984e65263ac7348a\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.17.0?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"4.17.0\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}}, \"data\": {\"name\": \"data-eb3ee5c90b71e8cc55c3c1d229156cbe\"}, \"mark\": \"bar\", \"encoding\": {\"color\": {\"field\": \"type\", \"type\": \"nominal\"}, \"tooltip\": [{\"field\": \"amount\", \"title\": \"Number of reads aligned\", \"type\": \"quantitative\"}, {\"field\": \"type\", \"type\": \"nominal\"}], \"x\": {\"aggregate\": \"sum\", \"axis\": {\"format\": \"%\"}, \"field\": \"amount\", \"stack\": \"normalize\", \"title\": \"Percent\", \"type\": \"quantitative\"}}, \"height\": 400, \"title\": \"Reads aligned to the human reference genome\", \"width\": 700, \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.17.0.json\", \"datasets\": {\"data-eb3ee5c90b71e8cc55c3c1d229156cbe\": [{\"amount\": 193250, \"type\": \"unaligned\"}, {\"amount\": 8177, \"type\": \"aligned\"}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from plotting import bowtie2_alignment_plot\n",
    "bowtie2_alignment_plot.plot_alignment(log)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e1d628ff-f0d5-4324-b844-dc9217a3ac7b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(201427, 4.06)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from utils import parse_bowtielog\n",
    "\n",
    "parse_bowtielog.parse_alignments(log)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "566da2d8-18f7-445c-9074-4dc89382be8d",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'alt' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [6]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0m cale\u001b[38;5;241m=\u001b[39m\u001b[43malt\u001b[49m\u001b[38;5;241m.\u001b[39mScale(scheme\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdark2\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'alt' is not defined"
     ]
    }
   ],
   "source": [
    "<style>\n",
    "\n",
    "        header {{\n",
    "        background-color: #333;\n",
    "        color: white;\n",
    "        padding: 20px;\n",
    "        text-align: center;\n",
    "        }}\n",
    "\n",
    "        h1 {{\n",
    "        color: #2ecc71;\n",
    "        }}\n",
    "\n",
    "        h2 {{\n",
    "        color: #3498db;\n",
    "        }}\n",
    "\n",
    "        h3 {{\n",
    "        color: #9b59b6;\n",
    "        }}\n",
    "\n",
    "        h4 {{\n",
    "        color: #f1c40f;\n",
    "        }}\n",
    "\n",
    "        div {{\n",
    "        color: #e67e22;\n",
    "        }}\n",
    "\n",
    "</style>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64cd1ca7-3e25-49d5-80aa-f5479ad2145d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Information about kraken and kaiju\n",
    "info = \"\"\"<div style=\"border: 1px solid black; padding: 10px; margin: 10px; width: 600px; border-radius: 10px; box-shadow: 5px 5px 5px grey; background-color: #cccccc;\">\n",
    "            <h1>Kraken2 and Kaiju</h1>\n",
    "            <ul>\n",
    "            <li>Kraken2 and Kaiju are tools for identifying the taxonomic origin of <strong>DNA sequences</strong></li>\n",
    "            <li>They use <strong>hierarchical classification systems</strong> and <strong>databases of pre-computed k-mer hashes</strong> to classify query sequences</li>\n",
    "            <li>Both programs may be used to <strong>rapidly and accurately classify large numbers of sequences</strong>, even in the absence of significant sequence similarity to known reference sequences</li>\n",
    "            </ul>\n",
    "        </div>\n",
    "        \"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a13389e1-3605-4bc8-9742-9a5472eddcab",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf_and_svg = \"\"\"\n",
    "\n",
    "            <h1> \n",
    "                This is a PDF \n",
    "            </h1>\n",
    "            <!-- <iframe src=\"{{pdf}}#toolbar=0\" width=\"100%\" height=\"900px\" border=\"0\"> -->\n",
    "            </iframe>\n",
    "            \n",
    "            <h1 style=\"color:orange; text-align:center; font-style:italic;\">\n",
    "                This is an embedded SVG\n",
    "            </h1>\n",
    "            <div style=\"display:flex;justify-content:center;align-items:center;width:100%;height:100%;margin:40px;\">\n",
    "                {svg}\n",
    "            </div>\n",
    "\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "36fe884a-092f-46af-b171-e3ceb3f273c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from bs4 import BeautifulSoup\n",
    "import re\n",
    "\n",
    "summary_information = {}\n",
    "\n",
    "with open(\"../virusclassification_nextflow/results/sample09_S8/fastp/sample09_S8.fastp.html\", \"r\") as f:\n",
    "    soup = BeautifulSoup(f, \"html.parser\")\n",
    "\n",
    "for table in soup.find_all(\"table\", class_=\"summary_table\")[:4]:\n",
    "    for row in table.find_all(\"tr\"):\n",
    "        cells = row.find_all(\"td\")\n",
    "        key = cells[0].text\n",
    "        value = cells[1].text\n",
    "        summary_information[key] = value\n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "f9be245d-db2d-46e2-b103-a2fa6195ae30",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'fastp version:': '0.23.2 (https://github.com/OpenGene/fastp)',\n",
       " 'sequencing:': 'paired end (76 cycles + 76 cycles)',\n",
       " 'mean length before filtering:': '74bp, 74bp',\n",
       " 'mean length after filtering:': '73bp, 73bp',\n",
       " 'duplication rate:': '2.756835%',\n",
       " 'Insert size peak:': '110',\n",
       " 'total reads:': '402.854000 K',\n",
       " 'total bases:': '29.808448 M',\n",
       " 'Q20 bases:': '28.459831 M (95.475722%)',\n",
       " 'Q30 bases:': '25.905512 M (86.906611%)',\n",
       " 'GC content:': '61.962052%',\n",
       " 'reads passed filters:': '402.854000 K (99.125509%)',\n",
       " 'reads with low quality:': '3.382000 K (0.832169%)',\n",
       " 'reads with too many N:': '22 (0.005413%)',\n",
       " 'reads too short:': '150 (0.036909%)'}"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ac1cb132-52e0-4aff-b7c0-aad8dceae965",
   "metadata": {},
   "outputs": [],
   "source": [
    "from utils import parse_fastp_report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "211c54c2-867e-4831-af6b-f51e01dabe2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "test = parse_fastp_report.parse_fastp(\"../virusclassification_nextflow/results/sample09_S8/fastp/sample09_S8.fastp.html\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a922aa87-294a-4520-bc50-eb05694954ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'<table border=\"1\" class=\"dataframe\">\\n  <thead>\\n    <tr style=\"text-align: right;\">\\n      <th></th>\\n      <th>description</th>\\n      <th>value</th>\\n    </tr>\\n  </thead>\\n  <tbody>\\n    <tr>\\n      <th>0</th>\\n      <td>fastp version:</td>\\n      <td>0.23.2 (https://github.com/OpenGene/fastp)</td>\\n    </tr>\\n    <tr>\\n      <th>1</th>\\n      <td>sequencing:</td>\\n      <td>paired end (76 cycles + 76 cycles)</td>\\n    </tr>\\n    <tr>\\n      <th>2</th>\\n      <td>mean length before filtering:</td>\\n      <td>74bp, 74bp</td>\\n    </tr>\\n    <tr>\\n      <th>3</th>\\n      <td>mean length after filtering:</td>\\n      <td>73bp, 73bp</td>\\n    </tr>\\n    <tr>\\n      <th>4</th>\\n      <td>duplication rate:</td>\\n      <td>2.756835%</td>\\n    </tr>\\n    <tr>\\n      <th>5</th>\\n      <td>Insert size peak:</td>\\n      <td>110</td>\\n    </tr>\\n    <tr>\\n      <th>6</th>\\n      <td>total reads:</td>\\n      <td>402.854000 K</td>\\n    </tr>\\n    <tr>\\n      <th>7</th>\\n      <td>total bases:</td>\\n      <td>29.808448 M</td>\\n    </tr>\\n    <tr>\\n      <th>8</th>\\n      <td>Q20 bases:</td>\\n      <td>28.459831 M (95.475722%)</td>\\n    </tr>\\n    <tr>\\n      <th>9</th>\\n      <td>Q30 bases:</td>\\n      <td>25.905512 M (86.906611%)</td>\\n    </tr>\\n    <tr>\\n      <th>10</th>\\n      <td>GC content:</td>\\n      <td>61.962052%</td>\\n    </tr>\\n    <tr>\\n      <th>11</th>\\n      <td>reads passed filters:</td>\\n      <td>402.854000 K (99.125509%)</td>\\n    </tr>\\n    <tr>\\n      <th>12</th>\\n      <td>reads with low quality:</td>\\n      <td>3.382000 K (0.832169%)</td>\\n    </tr>\\n    <tr>\\n      <th>13</th>\\n      <td>reads with too many N:</td>\\n      <td>22 (0.005413%)</td>\\n    </tr>\\n    <tr>\\n      <th>14</th>\\n      <td>reads too short:</td>\\n      <td>150 (0.036909%)</td>\\n    </tr>\\n  </tbody>\\n</table>'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test.to_html()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "3fe22365-2f01-400c-bea1-dc355cc078ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>fastp version:</td>\n",
       "      <td>0.23.2 (https://github.com/OpenGene/fastp)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sequencing:</td>\n",
       "      <td>paired end (76 cycles + 76 cycles)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mean length before filtering:</td>\n",
       "      <td>74bp, 74bp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mean length after filtering:</td>\n",
       "      <td>73bp, 73bp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>duplication rate:</td>\n",
       "      <td>2.756835%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Insert size peak:</td>\n",
       "      <td>110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total reads:</td>\n",
       "      <td>402.854000 K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>total bases:</td>\n",
       "      <td>29.808448 M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Q20 bases:</td>\n",
       "      <td>28.459831 M (95.475722%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Q30 bases:</td>\n",
       "      <td>25.905512 M (86.906611%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GC content:</td>\n",
       "      <td>61.962052%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>reads passed filters:</td>\n",
       "      <td>402.854000 K (99.125509%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>reads with low quality:</td>\n",
       "      <td>3.382000 K (0.832169%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>reads with too many N:</td>\n",
       "      <td>22 (0.005413%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>reads too short:</td>\n",
       "      <td>150 (0.036909%)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      description                                       value\n",
       "0                  fastp version:  0.23.2 (https://github.com/OpenGene/fastp)\n",
       "1                     sequencing:          paired end (76 cycles + 76 cycles)\n",
       "2   mean length before filtering:                                  74bp, 74bp\n",
       "3    mean length after filtering:                                  73bp, 73bp\n",
       "4               duplication rate:                                   2.756835%\n",
       "5               Insert size peak:                                         110\n",
       "6                    total reads:                                402.854000 K\n",
       "7                    total bases:                                 29.808448 M\n",
       "8                      Q20 bases:                    28.459831 M (95.475722%)\n",
       "9                      Q30 bases:                    25.905512 M (86.906611%)\n",
       "10                    GC content:                                  61.962052%\n",
       "11          reads passed filters:                   402.854000 K (99.125509%)\n",
       "12        reads with low quality:                      3.382000 K (0.832169%)\n",
       "13         reads with too many N:                              22 (0.005413%)\n",
       "14               reads too short:                             150 (0.036909%)"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
