
    <!DOCTYPE html>
    <html>
        <head>
            <title>Report of sample11_S6 </title>
            <script src="https://cdn.jsdelivr.net/npm/vega@5"></script>
            <script src="https://cdn.jsdelivr.net/npm/vega-lite@4"></script>
            <script src="https://cdn.jsdelivr.net/npm/vega-embed@6"></script>
            
            <style>
                body {
                font-family: Arial, sans-serif;
                font-size: 16px;
                line-height: 1.5;
                }
                header {
                background: linear-gradient(to right, #000000, #333333);
                color: white;
                padding: 40px;
                text-align: center;
                box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19);
                }
                h1 {
                font-size: 3em;
                margin: 0;
                text-align: center;
                }
                h2 {
                font-size: 2em;
                margin-bottom:30px;
                text-align: center;
                }
                section {
                margin: 80px 0;
                }
                hr {
                border: 0;
                height: 1px;
                background: #333;
                background-image: linear-gradient(to right, #ccc, #333, #ccc);
                }
                table {
                border-collapse: collapse;
                }
                th, td {
                border: 1px solid #ccc;
                padding: 10px;
                text-align: left;
                }
                th {
                background-color: #eee;
                }
                
            </style>

        </head>
        
        <body>
        
            <header>
                <h1>
                Pandemic Prepardeness Report 
                </h1>
                <h2>
                Report of sample11_S6
                </h2>
            </header>
            
            <h2>
                Information about Reads
            </h2>
            <table style="margin-bottom:30px;">
                <tr>
                    <th>Total reads</th>
                    <th>Reads aligned to human genome</th>
                    <th>Reads NOT aligned to human genome</th>
                </tr>
                <tr>
                    <td>16,684</td>
                    <td>15,367 (92.11%)</td>
                    <td>1,317 (7.89%)</td>
                </tr>
            </table>

            <div id="aligned" style="display:flex;width:100%;height:100%;"></div>
            <script type="text/javascript">
                vegaEmbed("#aligned", {
  "$schema": "https://vega.github.io/schema/vega-lite/v4.17.0.json",
  "config": {
    "view": {
      "continuousHeight": 300,
      "continuousWidth": 400
    }
  },
  "data": {
    "name": "data-62194969c72917a0f10041a474973727"
  },
  "datasets": {
    "data-62194969c72917a0f10041a474973727": [
      {
        "amount": 1317,
        "type": "unaligned"
      },
      {
        "amount": 15367,
        "type": "aligned"
      }
    ]
  },
  "encoding": {
    "color": {
      "field": "type",
      "scale": {
        "scheme": "dark2"
      },
      "type": "nominal"
    },
    "tooltip": [
      {
        "field": "amount",
        "title": "Number of reads aligned",
        "type": "quantitative"
      },
      {
        "field": "type",
        "type": "nominal"
      }
    ],
    "x": {
      "aggregate": "sum",
      "axis": {
        "format": "%"
      },
      "field": "amount",
      "stack": "normalize",
      "title": "Percent",
      "type": "quantitative"
    }
  },
  "mark": "bar",
  "title": "Reads aligned to the human reference genome"
});
            </script>
            
            <hr />
            
            <!-- KRAKEN RAW -->
            <h2> 
            Raw reads classfied with Kraken and Kaiju
            </h2>
            
            <div style="border: 1px solid black; padding: 10px; margin: 10px; width: 1000px; border-radius: 10px; box-shadow: 5px 5px 5px grey; background-color: #cccccc; display:flex;justify-content:center;">
            <ul>
            <li>KRAKEN2 and KAIJU are both popular bioinformatics tools for accurately and efficiently identifying and categorizing large numbers of short DNA sequences.</li>
            <li>KRAKEN2 is based on a probabilistic model and uses a pre-constructed database of genomic sequences, while KAIJU uses a database of reference sequences and an efficient indexing method.</li>
            <li>KRAKEN2 and KAIJU have been shown to produce highly accurate results in a short amount of time, but KRAKEN2 may require more computational resources.</li>
            <li>Both KRAKEN2 and KAIJU are open-source and freely available for use, making them valuable resources for researchers in the field of genomics.</li>
            </ul>
            </div>

            <h3>
            Kraken classification
            </h3>
            <div id="kraken_raw" style="display:flex;justify-content:center;align-items:center;width:100%;height:100%;margin:40px;"></div>
            <script type="text/javascript">
                vegaEmbed("#kraken_raw", {
  "$schema": "https://vega.github.io/schema/vega-lite/v4.17.0.json",
  "config": {
    "view": {
      "continuousHeight": 300,
      "continuousWidth": 400
    }
  },
  "datasets": {
    "data-b0c4e0d3bc9885e34157763d54e37ead": [
      {
        "domain": "Virus",
        "level": "S",
        "name": "                    Human alphaherpesvirus 1",
        "new_est_reads": 259,
        "percent": 0.1236999999999999,
        "tax_id": 10298
      }
    ],
    "data-b1aa583138a0f21f1a34d00dd7b7dfd5": [
      {
        "domain": "Eukaryota",
        "level": "D",
        "name": "    Eukaryota",
        "new_est_reads": 1815,
        "percent": 0.8639,
        "tax_id": 2759
      },
      {
        "domain": "Virus",
        "level": "D",
        "name": "  Viruses",
        "new_est_reads": 259,
        "percent": 0.1236999999999999,
        "tax_id": 10239
      },
      {
        "domain": "Bacteria",
        "level": "D",
        "name": "    Bacteria",
        "new_est_reads": 26,
        "percent": 0.0124,
        "tax_id": 2
      }
    ]
  },
  "hconcat": [
    {
      "data": {
        "name": "data-b0c4e0d3bc9885e34157763d54e37ead"
      },
      "encoding": {
        "color": {
          "field": "name",
          "title": null,
          "type": "nominal"
        },
        "tooltip": [
          {
            "field": "domain",
            "type": "nominal"
          },
          {
            "field": "new_est_reads",
            "title": "Number of reads",
            "type": "quantitative"
          }
        ],
        "x": {
          "axis": {
            "format": ".1%"
          },
          "field": "percent",
          "title": "Percent of reads",
          "type": "quantitative"
        },
        "y": {
          "field": "name",
          "sort": "-x",
          "title": null,
          "type": "nominal"
        }
      },
      "height": 500,
      "mark": "bar",
      "title": "Kraken classification raw",
      "width": 500
    },
    {
      "data": {
        "name": "data-b1aa583138a0f21f1a34d00dd7b7dfd5"
      },
      "encoding": {
        "color": {
          "field": "name",
          "title": null,
          "type": "nominal"
        },
        "tooltip": [
          {
            "field": "domain",
            "type": "nominal"
          },
          {
            "field": "new_est_reads",
            "title": "Number of reads",
            "type": "quantitative"
          }
        ],
        "x": {
          "axis": {
            "format": ".1%"
          },
          "field": "percent",
          "title": "Percent of reads",
          "type": "quantitative"
        },
        "y": {
          "field": "name",
          "sort": "-x",
          "title": null,
          "type": "nominal"
        }
      },
      "height": 500,
      "mark": "bar",
      "title": "Kraken classification raw",
      "width": 500
    }
  ],
  "resolve": {
    "scale": {
      "color": "independent"
    }
  }
});
            </script>
            
            <!-- KAIJU RAW -->
            <h3>
            Kaiju classification
            </h3>
            <div id="kaiju_raw" style="display:flex;justify-content:center;align-items:center;width:100%;height:100%;margin:40px;"></div>
            <script type="text/javascript">
                vegaEmbed("#kaiju_raw", {
  "$schema": "https://vega.github.io/schema/vega-lite/v4.17.0.json",
  "config": {
    "view": {
      "continuousHeight": 300,
      "continuousWidth": 400
    }
  },
  "data": {
    "name": "data-5c5887267d2503c7bbab9478104daf29"
  },
  "datasets": {
    "data-5c5887267d2503c7bbab9478104daf29": [
      {
        "percent": 0.05623902,
        "reads": 128,
        "taxon_id": 10298,
        "taxon_name": "Human alphaherpesvirus 1",
        "taxonomy": [
          "Viruses",
          "Duplodnaviria",
          "Human alphaherpesvirus 1",
          "Simplexvirus",
          "Alphaherpesvirinae",
          "Herpesviridae",
          "Herpesvirales",
          "Herviviricetes",
          "Peploviricota",
          "Heunggongvirae"
        ]
      },
      {
        "percent": 0.01669596,
        "reads": 38,
        "taxon_id": 10294,
        "taxon_name": "Simplexvirus",
        "taxonomy": [
          "Herpesvirales",
          "Peploviricota",
          "Herviviricetes",
          "Viruses",
          "Herpesviridae",
          "Alphaherpesvirinae",
          "Duplodnaviria",
          "Simplexvirus",
          "Heunggongvirae"
        ]
      }
    ]
  },
  "encoding": {
    "color": {
      "field": "taxon_name",
      "title": null,
      "type": "nominal"
    },
    "tooltip": [
      {
        "field": "taxonomy",
        "type": "ordinal"
      },
      {
        "field": "reads",
        "title": "Number of reads",
        "type": "quantitative"
      }
    ],
    "x": {
      "axis": {
        "format": ".1%"
      },
      "field": "percent",
      "title": "Percent of reads",
      "type": "quantitative"
    },
    "y": {
      "field": "taxon_name",
      "sort": "-x",
      "title": null,
      "type": "nominal"
    }
  },
  "height": 500,
  "mark": "bar",
  "title": "Kaiju classification raw",
  "width": 500
});
            </script>
            
            <hr />
            
            <h2>
            Contig information
            </h2>
            <h3>
            Histogram of megahit contigs
            </h3>
            <div style="border: 1px solid black; padding: 10px; margin: 10px; width: 1000px; border-radius: 10px; box-shadow: 5px 5px 5px grey; background-color: #cccccc; display:flex;justify-content:center;">
            <ul>
            <li>MEGAHIT is a popular and highly-efficient bioinformatics tool for assembling large and complex genomes.</li>
            <li>It is based on a De Bruijn graph approach and is capable of accurately assembling millions of reads into a complete genome in a short amount of time.</li>
            <li>MEGAHIT has been widely used in a variety of genomic studies and has been shown to produce high-quality assemblies that are comparable to those generated by other leading assemblers.</li>
            </ul>
            </div>


            <!-- PLOT MEGAHIT -->
            
            <div id="megahit_histo" style="display:flex;justify-content:center;align-items:center;width:100%;height:100%;margin:40px;"></div>
            <script type="text/javascript">
                vegaEmbed("#megahit_histo", {
  "$schema": "https://vega.github.io/schema/vega-lite/v4.17.0.json",
  "config": {
    "view": {
      "continuousHeight": 300,
      "continuousWidth": 400
    }
  },
  "data": {
    "name": "data-9999e5271c0f64a960914991fb68f099"
  },
  "datasets": {
    "data-9999e5271c0f64a960914991fb68f099": [
      {
        "length": 509,
        "name": "k79_13",
        "sequence": "AGACCAGGCCAGCAGGAGATGGAACCTTCCAGAAGTGGGCAGCTGTGGTGGTGCCTTCTGGAGAAGAGCAGAGATACACATGCCATGTACAGCATGAGGGGCTGCCGAAGCCCCTCACCCTGAGATGGGAGCCATCTTCCCAGTCCACCATCCCCATCGTGGGCATTGTTGCTGGCCTGGCTGTCCTAGCAGTTGTGGTCATCGGAGCTGTGGTCGCTACTGTGATGTGTAGGAGGAAGAGCTCAGGTGGAAAAGGAGGGAGCTACTCTCAGGCTGCGTCCAGCGACAGTGCCCAGGGCTCTGATGTGTCTCTCACAGCTTGAAAAGCCTGAGACAGCTGTCTTGTGAGGGACTGAGATGCAGGATTTCTTCACTCCTCCCCTTTGTGACTTCAAGAGCCTCTGGCATCTCTTTCTGCAAAGGCATCTGAATGTGTCTGCGTTCCTGTTAGCATAATGTGAGGAGGTGGAGAGACAGCCCACCCCCGTGTCCACTGTGACCCCTGTTCC"
      },
      {
        "length": 370,
        "name": "k79_7",
        "sequence": "GTGGAGCCACTCCACGCATGTGTCTTCCAGGTAGGCTCTCTGGTGCTCCGCCTCACGGGCAGCGGAGAGCCTACCTGGAGGGCACGTGCGTGGAGTGGCTCCGCAGATACCTGGAGAACGGGAAGGAGACGCTGCAGCGCGCGGAACACCCAAAGACACACGTGACCCACCATCCCGTCTCTGACCATGAGGCCACCCTGAGGTGCTGGGCCCTGGGCTTCTACCCTGCGGAGATCACACTGACCTGGCAGTGGGATGGGGAGGACCAAACTCAGGACACCGAGCTTGTGGAGACCAGGCCAGCAGGAGATGGAACCTTCCAGAAGTGGGCAGCTGTGGTGGTGCCTTCTGGAGAAGAGCAGAGATACAC"
      },
      {
        "length": 341,
        "name": "k79_12",
        "sequence": "TCTTCTTCCCCTTCTTTTCTTCCCCTTCCCCTTCTTTTCAATTTCTCCCACGTCAAGAGAAAAGCCAACATGTTTTTCCTCCAATGCATAAAAGGAACTTCCATAGGGCTGGCAGGAGTCAGGCTGTTCAAGACAACTGGAAGGAGTTGAATAACATCTATCCAGTGAGTCCTGCAAGACTTCAGGCCCTTTCTCATCCAGCAGCTCCCTGCTGAGCCTGGGGCATGGTGGGTTTTGATCTTCTTCCCCTTCTTTTCTTCCCCTTCTTCTTTCCTTCTTTGATCTTCTTCCCCTTCTTTTCTTCCCCTTCCCCTTCTTTTCAATTTCTCCCACGTCAAGAG"
      },
      {
        "length": 336,
        "name": "k79_9",
        "sequence": "TTTTTCAAAGTAAACGCTTCGGGCCCCGCGGGACACTCAGCTAAGAGCATCGAGGGGGCCCGAAGCGTTTACTTTGAAAAAATTAGAGTGTTCAAAGCAGGCCCGAGCCGCCTGGATACCGCAGCTAGGAATAATGGAATAGGACCGCGGTTCTATTTTGTTGGTTTTCGGAACTGAGGCCATGATTAAGAGGGACGGCCGGGGGCATTCGTATTGCGCCGCTAGAGGTGAAATTCTTGGACCGGCGCAAGACGGACCAGAGCGAAAGCATTTGCCAAGAATGTTTTCATTAATCAAGAACGAAAGTCGGAGGTTCGAAGACGATCAGATACCGTC"
      },
      {
        "length": 306,
        "name": "k79_4",
        "sequence": "AGACCAGGCCAGCAGGAGATGGAACCTTCCAGAAGTGGGCAGCTGTGGTGGTGCCTTCTGGAGAAGAGCAGAGATACACGTGCCATGTTCAGCACGAGGGGCTGCCGGAGCCCCTCACCCTGAGATGGAAGCCGTCTTCCCAGCCCACCATCCCCATCGTGGGCATCGTTGCTGGCCTGGCTGTCCTGGCTGTCCTAGCTGTCCTAGGAGCTATGGTGGCTGTTGTGATGTGTAGGAGGAAGAGCTCAGGTGGAAAAGGAGGGAGCTGCTCTCAGGCTGCGTCCAGCAACAGTGCCCAGGGCTCTG"
      },
      {
        "length": 291,
        "name": "k79_8",
        "sequence": "CTCTCCTTTCAGGATTTGAGTGAAACACCATGTCAAGTCGATTTAACTACATCTTCTATAGACGGAGCATGGGGTTCAGGTGTTGGCTTTAATTTAGTTTGTACTGTTAGTTTAACAGAATGCCCAACTTTTTTAACTTCCATCAAGGCATGTGATTCTGCAATGTGTTATGGGTCTACCACAACCAACCTAATGAGAGGGCAGAACACAGTCCATATTGTTGGAAAGGGAGGTCATTCTGGTTCAAAATTCATGTGTTGTCATGACACAAAGTGCTCTAGTACAGGCCTG"
      },
      {
        "length": 280,
        "name": "k79_11",
        "sequence": "GTCGTAGGCGAACTGGTCATGCCCGCGGAGGAGGCGCCCGTCGGGCCCCAGGTCGCAGCCAAACATCCTCTGGAGGGTGTGAGACCCGGCCTCGCTCTGGTTGTAGTAGCCGCGCAGTTTCCGCAGGTTCACTCGGTCAGCCTGTGCATGGCGCCCCGAACCCTCATCCTGCTGCTCTCGGGAGCCCTGGCCCTGACCGAGACCTGGGCCGGCTCCCACTCCATGAGGTATTTCTACACCTCCGTGTCCCGGCCCGGCCGCGGGGAGCCCCGCTTCATCG"
      },
      {
        "length": 271,
        "name": "k79_3",
        "sequence": "GCACCACCACCCACGGAATCGAGAAAGAGCTATCAATCTGTCAATCCTGTCCGTGTGTGGTGGTGCATGGCCGTTCTTAGTTGGTGGAGCGATTTGTCTGGTTAATTCCGATAACGAACGAGACTCTGGCATGCTAACTAGTTACGCGACCCCCGAGCGGTCGGCGTCCCCCAACTTCTTAGAGGGACAAGTGGCGTTCAGCCACCCGAGATTGAGCAATAACAGGTCTGTGATGCCCTTAGATGTCCGGTTCTGCACGCGCGCTACACTG"
      },
      {
        "length": 262,
        "name": "k79_2",
        "sequence": "CCCTGTTAGGGGTCGCCATGTGACCCCCTGACTTTATATATACAGACCCCCAACACCTACACATGGCCCCTTTGACTCAGACGCGGGGCCCGGGGTCGCCGTGGGACCCCCCTGACTCATACACAGAGACACGCCCCCACAACAAACACACAGGGACCGGGGTCGCCGTGTTAGGGGGCGTGGTCCCCACTGACTCATACGCAGGGCCCCCTTACTCACACGCATCTAGGGGGGTGGGGAGGAGCCGCCCGCCATATTTGGG"
      },
      {
        "length": 260,
        "name": "k79_6",
        "sequence": "CGCAGTTTCACTGTACCGGCCGTGCGTACTTAGATATGCTTGTCTCAAAGATTAAGCCATGCATGTCTAAGTACGCACGGCCGGTACAGTGAAACTGCGAATGGCTCATTAAATCAGTTATGGTTCCTTTGGTCGCTCGCTCCTCTCCTACTTGGATAACTGTGGTAATTCTAGAGCTAATACATGCCGACGGGCGCTGACCCCCTTCGCGGGGGGGATGCGTGCATTTATCAGATCAAAACCAACCCGGTCAGCCCCTC"
      },
      {
        "length": 258,
        "name": "k79_1",
        "sequence": "TCTTCTTCCCCTTCTTTTCTTCCCCTTCCCCTTCTTTTCAATTTCTCCCACGTCAAGAGCCAAGCCAAGGTACTGTTCCTCCAATGAGTAAACAGCACTGCTGTAGGGCTGGCCTAAGTCAGGCAGTTCAAGATAACCTGAAGGAGTCGAATAACATCTATCCAGTGAGTCCTGCAAGACTTCAGGCTCTTTCTCATCCAGCAGCTCCCTGCTGAGCCTGGAGTCTTGCAGGACTCACTGGATAGATGTTATTCGACT"
      },
      {
        "length": 258,
        "name": "k79_5",
        "sequence": "CAGGGTCTTCTTTCCCCGCTGATTCCGCCAAGCCCGTTCCCTTGGCTGTGGTTTCGCTGGATAGTAGGTAGGGACAGTGGGAATCTCGTTCATCCATTCATGCGCGTCACTAATTAGATGACGAGGCATTTGGCTACCTTAAGAGAGTCATAGTTACTCCCGCCGTTTACCCGCGCTTCATTGAATTTCTTCACTTTGACATTCAGAGCACTGGGCAGAAATCACATCGCGAAGGCCCGCGGCGGGTGTTGACGCGAT"
      },
      {
        "length": 257,
        "name": "k79_10",
        "sequence": "TGCAGGACTCACTGGATAGATGTTATTCAACTCCTTCAGGTTGTCTTGAACTGACTGACTCATGCCAGCCCTACAGAAGTGCCTTTTACATATTGGAGCAACAGCGTGTTGGCTTGGCTGTTGACATGGATGAAATTGAAAAGTACAAAGAAGTGGAAGAAGACCAAGACCCATCATGCTCCAGGTAACTTTGAGCAATTATGGATGCTTAATTCTGTGTTGACACCTGGAGATGCCAGGTCCAGGGAAAACAAGAG"
      },
      {
        "length": 248,
        "name": "k79_0",
        "sequence": "ATTTTTCGTCACTACCTCCCCGGGTCGGGAGTGGGTAATTTGCGCGCCTGCTGCCTTCCTTGGATGTGGTAGCCGTTTCTCAGGCTCCCTCTCCGGAATCGAACCCTGATTCCCCGTCACCCGTGGTCACCATGGTAGGCACGGCGACTACCATCGAAAGTTGATAGGGCAGACGTTCGAATGGGTCGTCGCCGCCACGGGGGGCGGGCGCCGGCGGCTTTGGTGACTCTAGATAACCTCGGGCCGAT"
      }
    ]
  },
  "encoding": {
    "x": {
      "bin": {
        "step": 500
      },
      "field": "length",
      "title": "Length (nt)",
      "type": "quantitative"
    },
    "y": {
      "aggregate": "count",
      "field": "length",
      "title": "Number of contigs",
      "type": "quantitative"
    }
  },
  "height": 500,
  "mark": "bar",
  "title": "Megahit contigs size",
  "width": 500
});
            </script>
            
            <!-- PLOT Kaiju and Cat -->
            <h3>
            Contigs classified with Kaiju and CAT
            </h3>
            
            <div style="border: 1px solid black; padding: 10px; margin: 10px; width: 1000px; border-radius: 10px; box-shadow: 5px 5px 5px grey; background-color: #cccccc; display:flex;justify-content:center;">
            <ul>
            <li>CAT (Classification and Annotation Tool) is a popular bioinformatics tool that uses a combination of machine learning algorithms and a pre-constructed database of genomic sequences to classify and annotate DNA sequences.</li>
            <li>The BLASTP aspect of CAT involves using the Basic Local Alignment Search Tool (BLASTP) to compare a query sequence to the database of reference sequences and identify similar sequences.</li>
            <li>BLASTP is a widely-used algorithm for sequence alignment and is known for its speed and accuracy in identifying similar sequences.</li>
          </ul>
          </div>
            
            <div id="kaiju_and_cat" style="display:flex;justify-content:center;align-items:center;width:100%;height:100%;margin:40px;"></div>
            <script type="text/javascript">
                vegaEmbed("#kaiju_and_cat", {
  "$schema": "https://vega.github.io/schema/vega-lite/v4.17.0.json",
  "config": {
    "view": {
      "continuousHeight": 300,
      "continuousWidth": 400
    }
  },
  "datasets": {
    "data-77cdc72e53c1199fc5a2e07ceedf7007": [
      {
        "class": "Mammalia: 1.00",
        "kingdom_cat": "Eukaryota",
        "last_level_cat": "Homo sapiens: 1.00",
        "name": "k79_1",
        "order": "Primates: 1.00",
        "phylum": "Chordata: 1.00",
        "reason": "based on 1/1 ORFs",
        "second_level_cat": "Homo: 1.00",
        "third_level_cat": "Hominidae: 1.00"
      },
      {
        "class": "Gammaproteobacteria: 1.00",
        "kingdom_cat": "Bacteria",
        "last_level_cat": "Klebsiella pneumoniae: 1.00",
        "name": "k79_12",
        "order": "Enterobacterales: 1.00",
        "phylum": "Proteobacteria: 1.00",
        "reason": "based on 1/1 ORFs",
        "second_level_cat": "Klebsiella: 1.00",
        "third_level_cat": "Enterobacteriaceae: 1.00"
      }
    ],
    "data-b63800248fb6d86f7226071cce8c0595": [
      {
        "kingdom": "Viruses",
        "last_level": " Puumala orthohantavirus",
        "name": "k79_8",
        "second_level": " Orthohantavirus",
        "taxon_id": 1980486,
        "third_level": " Mammantavirinae"
      }
    ]
  },
  "hconcat": [
    {
      "data": {
        "name": "data-b63800248fb6d86f7226071cce8c0595"
      },
      "encoding": {
        "color": {
          "field": "last_level",
          "title": null,
          "type": "nominal"
        },
        "tooltip": [
          {
            "field": "second_level",
            "type": "nominal"
          },
          {
            "field": "third_level",
            "type": "nominal"
          }
        ],
        "x": {
          "aggregate": "count",
          "axis": {
            "format": ".0f",
            "values": [
              0.0,
              1.0,
              2.0,
              3.0,
              4.0,
              5.0,
              6.0,
              7.0,
              8.0,
              9.0,
              10.0,
              11.0,
              12.0,
              13.0,
              14.0,
              15.0,
              16.0,
              17.0,
              18.0,
              19.0,
              20.0,
              21.0,
              22.0,
              23.0,
              24.0,
              25.0,
              26.0,
              27.0,
              28.0,
              29.0,
              30.0,
              31.0,
              32.0,
              33.0,
              34.0,
              35.0,
              36.0,
              37.0,
              38.0,
              39.0,
              40.0,
              41.0,
              42.0,
              43.0,
              44.0,
              45.0,
              46.0,
              47.0,
              48.0,
              49.0,
              50.0,
              51.0,
              52.0,
              53.0,
              54.0,
              55.0,
              56.0,
              57.0,
              58.0,
              59.0,
              60.0,
              61.0,
              62.0,
              63.0,
              64.0,
              65.0,
              66.0,
              67.0,
              68.0,
              69.0,
              70.0,
              71.0,
              72.0,
              73.0,
              74.0,
              75.0,
              76.0,
              77.0,
              78.0,
              79.0,
              80.0,
              81.0,
              82.0,
              83.0,
              84.0,
              85.0,
              86.0,
              87.0,
              88.0,
              89.0,
              90.0,
              91.0,
              92.0,
              93.0,
              94.0,
              95.0,
              96.0,
              97.0,
              98.0,
              99.0,
              100.0,
              101.0,
              102.0,
              103.0,
              104.0,
              105.0,
              106.0,
              107.0,
              108.0,
              109.0,
              110.0,
              111.0,
              112.0,
              113.0,
              114.0,
              115.0,
              116.0,
              117.0,
              118.0,
              119.0,
              120.0,
              121.0,
              122.0,
              123.0,
              124.0,
              125.0,
              126.0,
              127.0,
              128.0,
              129.0,
              130.0,
              131.0,
              132.0,
              133.0,
              134.0,
              135.0,
              136.0,
              137.0,
              138.0,
              139.0,
              140.0,
              141.0,
              142.0,
              143.0,
              144.0,
              145.0,
              146.0,
              147.0,
              148.0,
              149.0,
              150.0,
              151.0,
              152.0,
              153.0,
              154.0,
              155.0,
              156.0,
              157.0,
              158.0,
              159.0,
              160.0,
              161.0,
              162.0,
              163.0,
              164.0,
              165.0,
              166.0,
              167.0,
              168.0,
              169.0,
              170.0,
              171.0,
              172.0,
              173.0,
              174.0,
              175.0,
              176.0,
              177.0,
              178.0,
              179.0,
              180.0,
              181.0,
              182.0,
              183.0,
              184.0,
              185.0,
              186.0,
              187.0,
              188.0,
              189.0,
              190.0,
              191.0,
              192.0,
              193.0,
              194.0,
              195.0,
              196.0,
              197.0,
              198.0,
              199.0
            ]
          },
          "field": "last_level",
          "title": "Number of occurancies",
          "type": "quantitative"
        },
        "y": {
          "field": "last_level",
          "sort": {
            "field": "last_level:N",
            "op": "count",
            "order": "descending"
          },
          "title": null,
          "type": "nominal"
        }
      },
      "height": 500,
      "mark": "bar",
      "title": "Kaiju classification on MEGAHIT contigs",
      "width": 500
    },
    {
      "data": {
        "name": "data-77cdc72e53c1199fc5a2e07ceedf7007"
      },
      "encoding": {
        "color": {
          "field": "last_level_cat",
          "title": null,
          "type": "nominal"
        },
        "tooltip": [
          {
            "field": "second_level_cat",
            "type": "nominal"
          },
          {
            "field": "third_level_cat",
            "type": "nominal"
          }
        ],
        "x": {
          "aggregate": "count",
          "axis": {
            "format": ".0f",
            "values": [
              0.0,
              1.0,
              2.0,
              3.0,
              4.0,
              5.0,
              6.0,
              7.0,
              8.0,
              9.0,
              10.0,
              11.0,
              12.0,
              13.0,
              14.0,
              15.0,
              16.0,
              17.0,
              18.0,
              19.0,
              20.0,
              21.0,
              22.0,
              23.0,
              24.0,
              25.0,
              26.0,
              27.0,
              28.0,
              29.0,
              30.0,
              31.0,
              32.0,
              33.0,
              34.0,
              35.0,
              36.0,
              37.0,
              38.0,
              39.0,
              40.0,
              41.0,
              42.0,
              43.0,
              44.0,
              45.0,
              46.0,
              47.0,
              48.0,
              49.0,
              50.0,
              51.0,
              52.0,
              53.0,
              54.0,
              55.0,
              56.0,
              57.0,
              58.0,
              59.0,
              60.0,
              61.0,
              62.0,
              63.0,
              64.0,
              65.0,
              66.0,
              67.0,
              68.0,
              69.0,
              70.0,
              71.0,
              72.0,
              73.0,
              74.0,
              75.0,
              76.0,
              77.0,
              78.0,
              79.0,
              80.0,
              81.0,
              82.0,
              83.0,
              84.0,
              85.0,
              86.0,
              87.0,
              88.0,
              89.0,
              90.0,
              91.0,
              92.0,
              93.0,
              94.0,
              95.0,
              96.0,
              97.0,
              98.0,
              99.0,
              100.0,
              101.0,
              102.0,
              103.0,
              104.0,
              105.0,
              106.0,
              107.0,
              108.0,
              109.0,
              110.0,
              111.0,
              112.0,
              113.0,
              114.0,
              115.0,
              116.0,
              117.0,
              118.0,
              119.0,
              120.0,
              121.0,
              122.0,
              123.0,
              124.0,
              125.0,
              126.0,
              127.0,
              128.0,
              129.0,
              130.0,
              131.0,
              132.0,
              133.0,
              134.0,
              135.0,
              136.0,
              137.0,
              138.0,
              139.0,
              140.0,
              141.0,
              142.0,
              143.0,
              144.0,
              145.0,
              146.0,
              147.0,
              148.0,
              149.0,
              150.0,
              151.0,
              152.0,
              153.0,
              154.0,
              155.0,
              156.0,
              157.0,
              158.0,
              159.0,
              160.0,
              161.0,
              162.0,
              163.0,
              164.0,
              165.0,
              166.0,
              167.0,
              168.0,
              169.0,
              170.0,
              171.0,
              172.0,
              173.0,
              174.0,
              175.0,
              176.0,
              177.0,
              178.0,
              179.0,
              180.0,
              181.0,
              182.0,
              183.0,
              184.0,
              185.0,
              186.0,
              187.0,
              188.0,
              189.0,
              190.0,
              191.0,
              192.0,
              193.0,
              194.0,
              195.0,
              196.0,
              197.0,
              198.0,
              199.0
            ]
          },
          "field": "last_level_cat",
          "title": "Number of occurancies",
          "type": "quantitative"
        },
        "y": {
          "field": "last_level_cat",
          "sort": {
            "field": "last_level_cat:N",
            "op": "count",
            "order": "descending"
          },
          "title": null,
          "type": "nominal"
        }
      },
      "height": 500,
      "mark": "bar",
      "title": "CAT classification on MEGAHIT contigs",
      "width": 500
    }
  ],
  "resolve": {
    "scale": {
      "color": "independent"
    }
  }
});
            </script>
            
            <h3>
            Table containing information about contigs
            </h3>
            <table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>taxon_id</th>
      <th>length</th>
      <th>last_level_kaiju</th>
      <th>last_level_cat</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>k79_13</td>
      <td>NaN</td>
      <td>509</td>
      <td>NaN</td>
      <td>no support</td>
    </tr>
    <tr>
      <th>1</th>
      <td>k79_7</td>
      <td>NaN</td>
      <td>370</td>
      <td>NaN</td>
      <td>no support</td>
    </tr>
    <tr>
      <th>2</th>
      <td>k79_12</td>
      <td>NaN</td>
      <td>341</td>
      <td>NaN</td>
      <td>Klebsiella pneumoniae: 1.00</td>
    </tr>
    <tr>
      <th>3</th>
      <td>k79_9</td>
      <td>NaN</td>
      <td>336</td>
      <td>NaN</td>
      <td>no support</td>
    </tr>
    <tr>
      <th>4</th>
      <td>k79_4</td>
      <td>NaN</td>
      <td>306</td>
      <td>NaN</td>
      <td>no support</td>
    </tr>
    <tr>
      <th>5</th>
      <td>k79_8</td>
      <td>1980486.0</td>
      <td>291</td>
      <td>Puumala orthohantavirus</td>
      <td>no support</td>
    </tr>
    <tr>
      <th>6</th>
      <td>k79_11</td>
      <td>NaN</td>
      <td>280</td>
      <td>NaN</td>
      <td>no support</td>
    </tr>
    <tr>
      <th>7</th>
      <td>k79_3</td>
      <td>NaN</td>
      <td>271</td>
      <td>NaN</td>
      <td>no support</td>
    </tr>
    <tr>
      <th>8</th>
      <td>k79_2</td>
      <td>NaN</td>
      <td>262</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>9</th>
      <td>k79_6</td>
      <td>NaN</td>
      <td>260</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
            <hr />
            
            
        </body>
    </html>
    
